porcupine/plugins/run/no_terminal.py

```diff
@@ -242,20 +239,84 @@ def stop(self, *, quitting: bool = False) -> None:
         if not quitting:
             get_tab_manager().event_generate("<<FileSystemChanged>>")
 
+    def write_to_stdin(self, data: bytes) -> None:
+        if self._shell_process is not None:
+            assert self._shell_process.stdin is not None
+            self._shell_process.stdin.write(data)
+            self._shell_process.stdin.flush()  # TODO: may block, maybe that isn't an issue in practice?
+
+    def close_stdin(self) -> None:
+        if self._shell_process is not None:
+            assert self._shell_process.stdin is not None
+            self._shell_process.stdin.close()
+
+
+# A python-edit-tracking text widget lets you know whether the text is currently
+# being edited through invoking a method in Python, such as .insert(). This lets
+# you distinguish between default key bindings and whatever your code does.
+class PyEditTrackingText(tkinter.Text):
+    @copy_type(tkinter.Text.__init__)
```

Comment:
I wonder which is better `copy_type` or `functools.wraps`.

Reply:
`copy_type` is better because `wraps` infers the type from what it's decorating. So it becomes full of `Any`.